src/catalog/table_catalog.cpp

```diff
@@ -10,16 +10,277 @@
+              index_object->index_name.c_str());
+    return false;
+  }
+
+  valid_index_objects = true;
+  index_objects.insert(std::make_pair(index_object->index_oid, index_object));
+  index_names.insert(std::make_pair(index_object->index_name, index_object));
+  return true;
+}
+
+/* @brief   evict index catalog object from cache
+ * @param   index_oid
+ * @return  true if index_oid is found and evicted; false if not found
+ */
+bool TableCatalogObject::EvictIndexObject(oid_t index_oid) {
+  // std::lock_guard<std::mutex> lock(index_cache_lock);
+  if (!valid_index_objects) return false;
+
+  // find index name from index name cache
+  auto it = index_objects.find(index_oid);
+  if (it == index_objects.end()) {
+    return false;  // index oid not found in cache
+  }
+
+  auto index_object = it->second;
+  PL_ASSERT(index_object);
+  index_objects.erase(it);
+  index_names.erase(index_object->index_name);
+  return true;
+}
+
+/* @brief   evict index catalog object from cache
+ * @param   index_name
+ * @return  true if index_name is found and evicted; false if not found
+ */
+bool TableCatalogObject::EvictIndexObject(const std::string &index_name) {
+  // std::lock_guard<std::mutex> lock(index_cache_lock);
+  if (!valid_index_objects) return false;
+
+  // find index name from index name cache
+  auto it = index_names.find(index_name);
+  if (it == index_names.end()) {
+    return false;  // index name not found in cache
+  }
+
+  auto index_object = it->second;
+  PL_ASSERT(index_object);
+  index_names.erase(it);
+  index_objects.erase(index_object->index_oid);
+  return true;
+}
+
+/* @brief   evict all index catalog objects from cache
+ */
+void TableCatalogObject::EvictAllIndexObjects() {
+  // std::lock_guard<std::mutex> lock(index_cache_lock);
+  index_objects.clear();
+  index_names.clear();
+  valid_index_objects = false;
+}
+
+/* @brief   get all index objects of this table into cache
+ * @return  map from index oid to cached index object
+ */
+std::unordered_map<oid_t, std::shared_ptr<IndexCatalogObject>>
+TableCatalogObject::GetIndexObjects(bool cached_only) {
+  // std::lock_guard<std::mutex> lock(index_cache_lock);
+  if (!valid_index_objects && !cached_only) {
+    // get index catalog objects from pg_index
+    valid_index_objects = true;
+    index_objects =
+        IndexCatalog::GetInstance()->GetIndexObjects(table_oid, txn);
+  }
+  return index_objects;
+}
+
+/* @brief   get index object with index oid from cache
+ * @param   index_oid
+ * @param   cached_only   if cached only, return nullptr on a cache miss
+ * @return  shared pointer to the cached index object, nullptr if not found
+ */
+std::shared_ptr<IndexCatalogObject> TableCatalogObject::GetIndexObject(
+    oid_t index_oid, bool cached_only) {
+  // std::lock_guard<std::mutex> lock(index_cache_lock);
+  GetIndexObjects(cached_only);  // fetch index objects in case we have not
+  auto it = index_objects.find(index_oid);
```

Comment:
the following 4 lines are being used in multiple places. Shouldnt we make a function out of these 4 lines and call that function everytime.